Modules/Scripted/DICOMLib/DICOMUtils.py

```diff
@@ -156,3 +156,180 @@ def __enter__(self):
+    numOfLoadedNodes = nodeCollection.GetNumberOfItems()-actualLoadedNodes[nodeType]
+    if numOfLoadedNodes != loadedNodes[nodeType]:
+      logging.error("Number of loaded %s nodes was %d, but %d was expected" % \
+        (nodeType, numOfLoadedNodes, loadedNodes[nodeType]) )
+      success = False
+
+  return success
+
+#------------------------------------------------------------------------------
+# Collect all series instance UIDs in a DICOM database (the Slicer one by default)
+# Useful to get list of just imported series UIDs, for example:
+# newSeriesUIDs = [x for x in seriesUIDsAfter if x not in seriesUIDsBefore]
+def allSeriesUIDsInDatabase(database=None):
+  if database is None:
+    database = slicer.dicomDatabase
+  dicomWidget = slicer.modules.dicom.widgetRepresentation().self()
+  allSeriesUIDs = []
+  for patient in database.patients():
+    studies = database.studiesForPatient(patient)
+    series = [database.seriesForStudy(study) for study in studies]
+    seriesUIDs = [uid for uidList in series for uid in uidList]
+    allSeriesUIDs.extend(seriesUIDs)
+  return allSeriesUIDs
+
+#------------------------------------------------------------------------------
+# Collect series instance UIDs belonging to a list of files
+def seriesUIDsForFiles(files):
+  seriesUIDs = set()
+  for file in files:
+    seriesUID = slicer.dicomDatabase.seriesForFile(file)
+    if seriesUID != '':
+      seriesUIDs.add(seriesUID)
+  return seriesUIDs
+
+#------------------------------------------------------------------------------
+class OnlineDICOMData:
+  """Context manager to conveniently load DICOM files downloaded zipped from the internet
+  """
+  def __init__(self, url, zipFilePath=None, dicomDataDir=None, expectedNumOfFiles=0, selectedPlugins={}, loadedNodes={}):
+    from time import gmtime, strftime
+    self.url = url
+    if zipFilePath is None:
+      fileName = strftime("%Y%m%d_%H%M%S_", gmtime()) + 'OnlineDICOMData.zip'
+      self.zipFilePath = slicer.app.temporaryPath + '/' + fileName
+    else:
+      self.zipFilePath = zipFilePath
+    if dicomDataDir is None:
+      directoryName = strftime("%Y%m%d_%H%M%S_", gmtime()) + 'OnlineDICOMData'
+      self.dicomDataDir = slicer.app.temporaryPath + '/' + directoryName
+    else:
+      self.dicomDataDir = dicomDataDir
+    self.expectedNumberOfUnzippedFiles = expectedNumOfFiles
+    self.selectedPlugins = selectedPlugins
+    self.loadedNodes = loadedNodes
+  def __enter__(self):
+    if downloadFile(self.url, self.zipFilePath):
+      if unzipFile(self.zipFilePath, self.dicomDataDir, self.expectedNumberOfUnzippedFiles):
+        import os
+        dicomFiles = slicer.util.getFilesInDirectory(self.dicomDataDir)
+        if importDicom(self.dicomDataDir):
+          seriesUIDs = seriesUIDsForFiles(dicomFiles)
+          return loadSeriesWithVerification(seriesUIDs, self.selectedPlugins, self.loadedNodes)
+    return False
+  def __exit__(self, type, value, traceback):
```

Comment:
If we don't do any cleanup then why this is implemented as a context manager?

Reply:
Because Jc requested it...

Reply:
Do you think any cleanup would be necessary? Re-using already downloaded and unzipped data is I think valuable, so deleting those wouldn't improve things I think

Reply:
Maybe I can add a flag for the context in case somebody wants to clean up. Do you think this is something that could be useful?

Reply:
> Re-using already downloaded and unzipped data is I think valuable

Agreed.

That said, cache should be centralized. Directory and database used for testing should be specific to the context and cleared up at the end.